File: detox/ios/DetoxTestRunner/Utilities/WebSocket/WebSocket.m
//
//  WebSocket.m
//  Detox
//
//  Created by Tal Kol on 6/16/16.
//  Copyright © 2016 Wix. All rights reserved.
//

#import "WebSocket.h"
@import SocketRocket;

//Legacy API for apps that use old SocketRocket.
@interface SRWebSocket ()

- (void)send:(id)data;
- (BOOL)sendString:(NSString *)string error:(NSError **)error NS_SWIFT_NAME(send(string:));

@end


DTX_CREATE_LOG(WebSocket);

@interface WebSocket() <SRWebSocketDelegate>

@property (nonatomic, retain) NSString *sessionId;
@property (nonatomic, retain) SRWebSocket *websocket;

@end


@implementation WebSocket

- (instancetype)init
{
	self = [super init];
	if(self)
	{
		_delegateQueue = dispatch_queue_create("com.wix.detoxTestRunner.webSocket", dispatch_queue_attr_make_with_autorelease_frequency(NULL, DISPATCH_AUTORELEASE_FREQUENCY_WORK_ITEM));
	}
	return self;
}

- (void)connectToServer:(NSString*)url withSessionId:(NSString*)sessionId
{
	if (self.websocket)
	{
		[self.websocket close];
		self.websocket = nil;
	}
	self.sessionId = sessionId;
	self.websocket = [[SRWebSocket alloc] initWithURL:[NSURL URLWithString:url]];
	self.websocket.delegateDispatchQueue = _delegateQueue;
	self.websocket.delegate = self;
	[self.websocket open];
}

- (void)close
{
	if (self.websocket)
	{
		[self.websocket close];
		self.websocket = nil;
	}
}

- (void)sendAction:(NSString*)type withParams:(NSDictionary*)params withMessageId:(NSNumber*)messageId
{
	NSDictionary *data = @{@"type": type, @"params": params, @"messageId": messageId};
	NSError *error;
	NSData *jsonData = [NSJSONSerialization dataWithJSONObject:data options:kNilOptions error:&error];
	if (jsonData == nil)
	{
		dtx_log_error(@"Error decoding sendAction encode - %@", error);
		return;
	}
	dtx_log_info(@"Action Sent: %@", type);
	NSString *json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
	
	[self.websocket sendString:json error:NULL];
}

- (void)receiveAction:(NSString*)json
{
	NSError *error;
	NSData *jsonData = [json dataUsingEncoding:NSUTF8StringEncoding];
	NSDictionary *data = [NSJSONSerialization JSONObjectWithData:jsonData options:kNilOptions error:&error];
	if (data == nil)
	{
		dtx_log_error(@"Error decoding receiveAction decode - %@", error);
		return;
	}
	NSString *type = [data objectForKey:@"type"];
	if (type == nil)
	{
		dtx_log_error(@"receiveAction missing type");
		return;
	}
	NSDictionary *params = [data objectForKey:@"params"];
	if (params != nil && ![params isKindOfClass:[NSDictionary class]])
	{
		dtx_log_error(@"receiveAction invalid params");
		return;
	}
	NSNumber *messageId = [data objectForKey:@"messageId"];
	if (messageId != nil && ![messageId isKindOfClass:[NSNumber class]])
	{
		dtx_log_error(@"receiveAction invalid messageId");
		return;
	}
	dtx_log_info(@"Action Received: %@", type);
//	dtx_log_debug(@"params: %@", params);
	[self.delegate webSocket:self didReceiveAction:type withParams:params withMessageId:messageId];
}

- (void)webSocketDidOpen:(SRWebSocket *)webSocket
{
	[self sendAction:@"login" withParams:@{@"sessionId": self.sessionId, @"role": @"testee"} withMessageId:@0];
	[self.delegate webSocketDidConnect:self];
}

- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessageWithString:(NSString *)string
{
	[self receiveAction:string];
}

- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error
{
	[self.delegate webSocket:self didFailWithError:error];
}

- (void)webSocket:(SRWebSocket *)webSocket didCloseWithCode:(NSInteger)code reason:(NSString *)reason wasClean:(BOOL)wasClean
{
	[self.delegate webSocket:self didCloseWithReason:reason];
}

@end

File: detox/detox/ios/Detox/Utilities/WebSocket.swift
//
//  WebSocket.swift
//  Detox
//
//  Created by Leo Natan (Wix) on 5/22/20.
//  Copyright © 2020 Wix. All rights reserved.
//

import Foundation

fileprivate let log = DetoxLog(category: "WebSocket")

protocol WebSocketDelegate: AnyObject {
	func webSocketDidConnect(_ webSocket: WebSocket)
	func webSocket(_ webSocket: WebSocket, didFailWith error: Error)
	func webSocket(_ webSocket: WebSocket, didReceiveAction type : String, params: [String: Any], messageId: NSNumber)
	func webSocket(_ webSocket: WebSocket, didCloseWith reason: String?)
}

class WebSocket : NSObject, URLSessionWebSocketDelegate {
	var sessionId: String?
	private var urlSession: URLSession!
	private var webSocketSessionTask: URLSessionWebSocketTask?
	weak var delegate: WebSocketDelegate?
	
	override init() {
		super.init()
		urlSession = URLSession(configuration: .default, delegate: self, delegateQueue: OperationQueue.main)
	}
	
	func connect(toServer server: URL, withSessionId sessionId: String) {
		self.sessionId = sessionId
		
		webSocketSessionTask = urlSession.webSocketTask(with: server)
		webSocketSessionTask?.resume()
	}
	
	func close() {
		webSocketSessionTask?.cancel(with: .normalClosure, reason: nil)
		webSocketSessionTask = nil
	}
	
	func sendAction(_ type: String, params: [String : Any], messageId: NSNumber) {
		let data : [String: Any] = ["type": type, "params": params, "messageId": messageId]
		do {
			let data = try JSONSerialization.data(withJSONObject: data, options: [])
			let message = URLSessionWebSocketTask.Message.data(data)
			webSocketSessionTask?.send(message) { error in
				if let error = error {
					log.error("Error sending message: \(error.localizedDescription)")
				}
			}
		} catch {
			log.error("Error encoding message: \(error.localizedDescription)")
		}
	}
	
	private func receive() {
		webSocketSessionTask?.receive { [weak self] result in
			switch result {
			case .failure(let error as NSError):
				log.error("Error receiving message: \(error.localizedDescription)")
			case .success(let message):
				switch message {
				case .string(let string):
					self?.receiveAction(json: string)
				case .data(let data):
					self?.receiveAction(json: String(data: data, encoding: .utf8)!)
				@unknown default:
					fatalError("Unknown websocket message type")
				}
				
				self?.receive()
			}
		}
	}
	
	func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol pr: String?) {
		onDidOpen()
		receive()
		
		delegate?.webSocketDidConnect(self)
	}
	
	
	func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
		let string: String?
		if let reason = reason, let str = String(data: reason, encoding: .utf8) {
			string = str
		} else {
			string = nil
		}
		
		delegate?.webSocket(self, didCloseWith: string)
	}
	
	func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
		if let error = error {
			delegate?.webSocket(self, didFailWith: error)
		} else {
			delegate?.webSocket(self, didCloseWith: nil)
		}
	}
	
	func receiveAction(json: String) {
		do {
			let jsonData = json.data(using: .utf8)!
			let obj = try JSONSerialization.jsonObject(with: jsonData, options: []) as! [String: Any]
			
			let type = obj["type"] as! String
			let params = obj["params"] as? [String: Any]
			let messageId = obj["messageId"] as! NSNumber
			
			log.info("Action received: \(type)")
			
			delegate?.webSocket(self, didReceiveAction: type, params: params ?? [:], messageId: messageId)
		} catch {
			log.error("Error decoding receiveAction decode: \(error.localizedDescription)")
		}
	}
	
	func onDidOpen() {
		sendAction("login", params: ["sessionId": sessionId!, "role": "app"], messageId: 0)
	}
}
